### ISSUE TITLE
Unable to set role permissions for new schema unless a new connection is created.

### ISSUE THREAD

**Issue by MialLewis** [apache/superset#29299]
### Bug description

I have a Trino database which is made available through `connection1`. If I create a new schema on this database, I cannot give users permission to access this schema via their `role`.

When navigating to `Settings` > `List Roles` > `Edit Record`, in the `Permissions` field I cannot see the new schema, only schema that existed before the creation of `connection1`.

If I now create a new connection, `connection2`, which is exactly the same as `connection1`, I can see the new schema,

This bug was discussed last year here: https://lists.apache.org/thread/bqogccdg3wdjkjsgytsqw1ylsmps756v

And now here: https://github.com/konzainc/konza-kube/issues/375

### How to reproduce the bug

1) Create a database connection to a Trino database, `test_connection_1` (other db's not tested).
2) On this database create a new schema `test_schema` through `SQL Lab`.
3) Navigate to `Settings` > `List Roles` > `Edit Record`. In the permissions field search for `test_schema`, which will not be found
4) Create a new database connection configured in the same way as the first, `test_connection_2`.
5) Repeat step 3, this time see `test_schema` is found through `test_connection_2`.

### Superset version

3.1.3

### Python version

3.10

### Node version

I don't know

### Browser

Firefox

### Checklist

- [X] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [X] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [X] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by Habeeb556** [comment]
Hello @MialLewis, I am facing the same issue. However, I worked around it by going to the database connections where you configure your database and clicking on `Test Connection"` This acts like a refresh instead of creating a new data source connection.

Additionally, I think this bug could be resolved by implementing an auto-refresh feature in Superset.

**Comment by MialLewis** [comment]
`Test Connection` didn't work to refresh schema for me, wonder if it depends on the connection type.

**Comment by Habeeb556** [comment]
@MialLewis What type of database connection did you use?

**Comment by algardas-sca** [comment]
Yeah, also had this problem - it's difficult to make superset refresh the schema_access permissions after updating the database schema. I have used this script as part of my provisioning setup:

```python
import superset
from superset import security_manager as sm

def refresh_db_schemas(dbname):
    from superset.daos.database import DatabaseDAO

    # superset does not automatically refresh the schema_access permissions after DB schema changes,
    # let's do it manually for now (code parts from upstream superset/superset/databases/api.py)
    database = DatabaseDAO.get_database_by_name(dbname)
    schemas = database.get_all_schema_names(cache=False)
    for schema in schemas:
        sm.add_permission_view_menu("schema_access", sm.get_schema_perm(database, schema)) 

app = superset.create_app()
with get_app().app_context():
  refresh_db_schemas("mydb")
```

**Comment by Habeeb556** [comment]
++ I encountered the same issue, and using `"Test Connection"` did not resolve it. The problem is related to role permissions. The schema name appeared in SQL Lab after refreshing the list from button, but the role permission issue persists with the new schema name not showing.

**Comment by rusackas** [comment]
Pinging the illustrious Trino review crew from the [rolodex](https://docs.google.com/spreadsheets/d/1U1qxiLvOX0kBTUGME1AHHi6Ywel6ECF8xk_Qy-V9R8c/edit?gid=0#gid=0):
@giftig
@sujiplr
@nytai
@cccs-tom
@villebro

**Comment by Habeeb556** [comment]
@dosu 

> Yeah, also had this problem - it's difficult to make superset refresh the schema_access permissions after updating the database schema. I have used this script as part of my provisioning setup:
> 
> ```python
> import superset
> from superset import security_manager as sm
> 
> def refresh_db_schemas(dbname):
>     from superset.daos.database import DatabaseDAO
> 
>     # superset does not automatically refresh the schema_access permissions after DB schema changes,
>     # let's do it manually for now (code parts from upstream superset/superset/databases/api.py)
>     database = DatabaseDAO.get_database_by_name(dbname)
>     schemas = database.get_all_schema_names(cache=False)
>     for schema in schemas:
>         sm.add_permission_view_menu("schema_access", sm.get_schema_perm(database, schema)) 
> 
> app = superset.create_app()
> with app.app_context():
>   refresh_db_schemas("mydb")
> ```

I believe the issue isn't just with forcing the retrieval of new schemas, but also with the role section being cached. For example, even after deleting and creating a new Datasource, the old Datasource still appears in `role_permission`.

**Comment by Habeeb556** [comment]
Hello @cccs-tom, I understand the workaround. However, I was referring to another issue where the old deleted database configuration is still cached. It appears when I open the list of roles and attempt to add `role permissions` for a specific schema.

**Comment by qleroy** [comment]
> Yeah, also had this problem - it's difficult to make superset refresh the schema_access permissions after updating the database schema. I have used this script as part of my provisioning setup:
> 
> import superset
> from superset import security_manager as sm
> 
> def refresh_db_schemas(dbname):
>     from superset.daos.database import DatabaseDAO
> 
>     # superset does not automatically refresh the schema_access permissions after DB schema changes,
>     # let's do it manually for now (code parts from upstream superset/superset/databases/api.py)
>     database = DatabaseDAO.get_database_by_name(dbname)
>     schemas = database.get_all_schema_names(cache=False)
>     for schema in schemas:
>         sm.add_permission_view_menu("schema_access", sm.get_schema_perm(database, schema)) 
> 
> app = superset.create_app()
> with app.app_context():
>   refresh_db_schemas("mydb")

Since SIP-95, `sm.get_schema_perm` gets a `catalog` parameter as a second parameter before `schema`, so we should use `sm.get_schema_perm(database_name, schema_name)`,
the correct snippet to update database schemas programmatically is

```python
import superset
from superset import security_manager as sm

def refresh_db_schemas(dbname):
    from superset.daos.database import DatabaseDAO

    # superset does not automatically refresh the schema_access permissions after DB schema changes,
    # let's do it manually for now (code parts from upstream superset/superset/databases/api.py)
    database = DatabaseDAO.get_database_by_name(dbname)
    schemas = database.get_all_schema_names(cache=False)
    for schema in schemas:
        sm.add_permission_view_menu("schema_access", sm.get_schema_perm(database, schema=schema)) 

app = superset.create_app()
with app.app_context():
  refresh_db_schemas("mydb")
```

**Comment by rusackas** [comment]
I presume the above comment was what's neede here, and this is safe to close out (since it's more of  a Q&A/help discussion) but let me know if that's not the case, and I'm happy to reopen!
